src/frontend/fcitx4frontend/fcitx4frontend.cpp

```diff
@@ -0,0 +1,358 @@
+/*
+ * SPDX-FileCopyrightText: 2020-2021 Vifly <viflythink@gmail.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ *
+ */
+
+#include "fcitx4frontend.h"
+#include <fstream>
+#include "fcitx-utils/dbus/message.h"
+#include "fcitx-utils/dbus/objectvtable.h"
+#include "fcitx-utils/dbus/servicewatcher.h"
+#include "fcitx-utils/log.h"
+#include "fcitx-utils/metastring.h"
+#include "fcitx-utils/standardpath.h"
+#include "fcitx/inputcontext.h"
+#include "fcitx/inputmethodentry.h"
+#include "fcitx/inputmethodmanager.h"
+#include "fcitx/instance.h"
+#include "fcitx/misc_p.h"
+#include "dbus_public.h"
+
+#define FCITX_INPUTMETHOD_DBUS_INTERFACE "org.fcitx.Fcitx.InputMethod"
+#define FCITX_INPUTCONTEXT_DBUS_INTERFACE "org.fcitx.Fcitx.InputContext"
+#define FCITX_DBUS_SERVICE "org.fcitx.Fcitx"
+
+namespace fcitx {
+
+namespace {
+
+std::vector<dbus::DBusStruct<std::string, int>>
+buildFormattedTextVector(const Text &text) {
+    std::vector<dbus::DBusStruct<std::string, int>> vector;
+    for (int i = 0, e = text.size(); i < e; i++) {
+        vector.emplace_back(std::make_tuple(
```

Comment:
While fcitx4's format flag is almost compatible:

There's one difference: (1<<3) bit means "no underline", while in fcitx 5 it means "underline". You need to flip this bit.

Reply:
See https://github.com/fcitx/fcitx/blob/38107d8454ec503dc9d93cf5f8e6e5ee7be88857/src/frontend/ipcportal/ipcportal.c#L714